// @clerk/fastify/dist/index.mjs
import {
  fastifyRequestToRequest
} from "./chunk-NY4ZCCUV.mjs";

// src/index.ts
export * from "@clerk/backend";

// src/clerkPlugin.ts
import fp from "fastify-plugin";

// src/types.ts
var ALLOWED_HOOKS = ["onRequest", "preHandler"];

// src/withClerkMiddleware.ts
import { AuthStatus } from "@clerk/backend/internal";

// src/clerkClient.ts
import { createClerkClient } from "@clerk/backend";

// src/constants.ts
import { constants } from "@clerk/backend/internal";
import { apiUrlFromPublishableKey } from "@clerk/shared/apiUrlFromPublishableKey";
var API_VERSION = process.env.CLERK_API_VERSION || "v1";
var SECRET_KEY = process.env.CLERK_SECRET_KEY || "";
var MACHINE_SECRET_KEY = process.env.CLERK_MACHINE_SECRET_KEY || "";
var PUBLISHABLE_KEY = process.env.CLERK_PUBLISHABLE_KEY || "";
var API_URL = process.env.CLERK_API_URL || apiUrlFromPublishableKey(PUBLISHABLE_KEY);
var JWT_KEY = process.env.CLERK_JWT_KEY || "";
var SDK_METADATA = {
  name: "@clerk/fastify",
  version: "2.6.31",
  environment: "production"
};
var { Cookies, Headers } = constants;

// src/clerkClient.ts
var clerkClient = createClerkClient({
  secretKey: SECRET_KEY,
  machineSecretKey: MACHINE_SECRET_KEY,
  apiUrl: API_URL,
  apiVersion: API_VERSION,
  jwtKey: JWT_KEY,
  userAgent: `${"@clerk/fastify"}@${"2.6.31"}`,
  sdkMetadata: SDK_METADATA
});

// src/withClerkMiddleware.ts
var withClerkMiddleware = (options) => {
  return async (fastifyRequest, reply) => {
    const req = fastifyRequestToRequest(fastifyRequest);
    const requestState = await clerkClient.authenticateRequest(req, {
      ...options,
      secretKey: options.secretKey || SECRET_KEY,
      publishableKey: options.publishableKey || PUBLISHABLE_KEY,
      acceptsToken: "any"
    });
    requestState.headers.forEach((value, key) => reply.header(key, value));
    const locationHeader = requestState.headers.get(Headers.Location);
    if (locationHeader) {
      return reply.code(307).send();
    } else if (requestState.status === AuthStatus.Handshake) {
      throw new Error("Clerk: handshake status without redirect");
    }
    fastifyRequest.auth = requestState.toAuth();
  };
};

// src/clerkPlugin.ts
var plugin = (instance, opts, done) => {
  instance.decorateRequest("auth", null);
  const hookName = opts.hookName || "preHandler";
  if (!ALLOWED_HOOKS.includes(hookName)) {
    throw new Error(`Unsupported hookName: ${hookName}`);
  }
  instance.addHook(hookName, withClerkMiddleware(opts));
  done();
};
var clerkPlugin = fp(plugin, {
  name: "@clerk/fastify",
  fastify: "5.x"
});

// src/getAuth.ts
import { getAuthObjectForAcceptedToken } from "@clerk/backend/internal";

// src/errors.ts
var createErrorMessage = (msg) => {
  return `\u{1F512} Clerk: ${msg.trim()}

For more info, check out the docs: https://clerk.com/docs,
or come say hi in our discord server: https://clerk.com/discord
`;
};
var pluginRegistrationRequired = createErrorMessage(`The "clerkPlugin" should be registered before using the "getAuth".
Example:

import { clerkPlugin } from '@clerk/fastify';

const server: FastifyInstance = Fastify({ logger: true });
server.register(clerkPlugin);
`);

// src/getAuth.ts
var getAuth = ((req, options) => {
  const authReq = req;
  if (!authReq.auth) {
    throw new Error(pluginRegistrationRequired);
  }
  return getAuthObjectForAcceptedToken({ authObject: authReq.auth, acceptsToken: options?.acceptsToken });
});
export {
  clerkClient,
  clerkPlugin,
  getAuth
};
