// @clerk/fastify/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  clerkClient: () => clerkClient,
  clerkPlugin: () => clerkPlugin,
  getAuth: () => getAuth
});
module.exports = __toCommonJS(index_exports);
__reExport(index_exports, require("@clerk/backend"), module.exports);

// src/clerkPlugin.ts
var import_fastify_plugin = __toESM(require("fastify-plugin"));

// src/types.ts
var ALLOWED_HOOKS = ["onRequest", "preHandler"];

// src/withClerkMiddleware.ts
var import_internal2 = require("@clerk/backend/internal");

// src/clerkClient.ts
var import_backend = require("@clerk/backend");

// src/constants.ts
var import_internal = require("@clerk/backend/internal");
var import_apiUrlFromPublishableKey = require("@clerk/shared/apiUrlFromPublishableKey");
var API_VERSION = process.env.CLERK_API_VERSION || "v1";
var SECRET_KEY = process.env.CLERK_SECRET_KEY || "";
var MACHINE_SECRET_KEY = process.env.CLERK_MACHINE_SECRET_KEY || "";
var PUBLISHABLE_KEY = process.env.CLERK_PUBLISHABLE_KEY || "";
var API_URL = process.env.CLERK_API_URL || (0, import_apiUrlFromPublishableKey.apiUrlFromPublishableKey)(PUBLISHABLE_KEY);
var JWT_KEY = process.env.CLERK_JWT_KEY || "";
var SDK_METADATA = {
  name: "@clerk/fastify",
  version: "2.6.31",
  environment: "production"
};
var { Cookies, Headers: Headers2 } = import_internal.constants;

// src/clerkClient.ts
var clerkClient = (0, import_backend.createClerkClient)({
  secretKey: SECRET_KEY,
  machineSecretKey: MACHINE_SECRET_KEY,
  apiUrl: API_URL,
  apiVersion: API_VERSION,
  jwtKey: JWT_KEY,
  userAgent: `${"@clerk/fastify"}@${"2.6.31"}`,
  sdkMetadata: SDK_METADATA
});

// src/utils.ts
var fastifyRequestToRequest = (req) => {
  const headers = new Headers(
    Object.keys(req.headers).reduce((acc, key) => {
      const value = req.headers[key];
      if (!value) {
        return acc;
      }
      if (typeof value === "string") {
        acc.set(key, value);
      } else {
        acc.set(key, value.join(","));
      }
      return acc;
    }, new Headers())
  );
  const dummyOriginReqUrl = new URL(req.url || "", `${req.protocol}://clerk-dummy`);
  return new Request(dummyOriginReqUrl, {
    method: req.method,
    headers
  });
};

// src/withClerkMiddleware.ts
var withClerkMiddleware = (options) => {
  return async (fastifyRequest, reply) => {
    const req = fastifyRequestToRequest(fastifyRequest);
    const requestState = await clerkClient.authenticateRequest(req, {
      ...options,
      secretKey: options.secretKey || SECRET_KEY,
      publishableKey: options.publishableKey || PUBLISHABLE_KEY,
      acceptsToken: "any"
    });
    requestState.headers.forEach((value, key) => reply.header(key, value));
    const locationHeader = requestState.headers.get(Headers2.Location);
    if (locationHeader) {
      return reply.code(307).send();
    } else if (requestState.status === import_internal2.AuthStatus.Handshake) {
      throw new Error("Clerk: handshake status without redirect");
    }
    fastifyRequest.auth = requestState.toAuth();
  };
};

// src/clerkPlugin.ts
var plugin = (instance, opts, done) => {
  instance.decorateRequest("auth", null);
  const hookName = opts.hookName || "preHandler";
  if (!ALLOWED_HOOKS.includes(hookName)) {
    throw new Error(`Unsupported hookName: ${hookName}`);
  }
  instance.addHook(hookName, withClerkMiddleware(opts));
  done();
};
var clerkPlugin = (0, import_fastify_plugin.default)(plugin, {
  name: "@clerk/fastify",
  fastify: "5.x"
});

// src/getAuth.ts
var import_internal3 = require("@clerk/backend/internal");

// src/errors.ts
var createErrorMessage = (msg) => {
  return `\u{1F512} Clerk: ${msg.trim()}

For more info, check out the docs: https://clerk.com/docs,
or come say hi in our discord server: https://clerk.com/discord
`;
};
var pluginRegistrationRequired = createErrorMessage(`The "clerkPlugin" should be registered before using the "getAuth".
Example:

import { clerkPlugin } from '@clerk/fastify';

const server: FastifyInstance = Fastify({ logger: true });
server.register(clerkPlugin);
`);

// src/getAuth.ts
var getAuth = ((req, options) => {
  const authReq = req;
  if (!authReq.auth) {
    throw new Error(pluginRegistrationRequired);
  }
  return (0, import_internal3.getAuthObjectForAcceptedToken)({ authObject: authReq.auth, acceptsToken: options?.acceptsToken });
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  clerkClient,
  clerkPlugin,
  getAuth,
  ...require("@clerk/backend")
});
